### ISSUE TITLE
PGML single quote behavior

### ISSUE THREAD

**Issue by Alex-Jordan** [openwebwork/pg#759]
In chasing an issue, I've built the following MWE:

```
DOCUMENT();

loadMacros(
  "PGstandard.pl",
  "PGML.pl",
);

TEXT(PGML(q!It's here.!));

BEGIN_TEXT
\{PGML(q!It's here too.!)\}
END_TEXT

BEGIN_PGML
[@ PGML(q!It's not here.!) @]*
END_PGML

ENDDOCUMENT();
```

This should print:
```
It's here.
It's here too.
It's not here.
```

But the last line is not printed. Something about the single quote, in a string argument to `PGML()`, being executed within a `[@...@]` block is not working.

Oddly, if you add a second single quote to the string then you will see the ouptut with two single quotes. For example
```
BEGIN_PGML
[@ PGML(q!It''s not here.!) @]*
END_PGML
```
will let you see `It''s not here.` in output. If you go up to 3 single quotes or any odd number of single quotes, it's gone again. But it faithfully shows the right number of single quotes when it is an even number.

**Comment by Alex-Jordan** [comment]
At first it looked like this is some edge case use from using `PGML::Format2` inside a `[@...@]` block. But I think it is a more significant issue with PGML parsing. Consider:

```
DOCUMENT();

loadMacros(
  "PGstandard.pl",
  "PGML.pl",
);

$content = q!it's a string!;

BEGIN_PGML
[$content]

[@$content@]

[@ q!it's a string! @]
END_PGML

ENDDOCUMENT();
```

I only see "it's a string" come out twice. I'm guessing that at some point, a string constructor like `q!...!` is not accounted for and something in the parsing expects that single quote to have a sibling.

**Comment by Alex-Jordan** [comment]
Maybe it's a problem that the contents of `[@...@]` is parsed at all, instead of just being executed?

**Comment by dpvc** [comment]
You are correct that there is parsing going on within the `[@ ... @]`.  In particular, quotes are expected to be balanced, so that you can do something like `[@ "@]" @]` and not match the quoted `@]`.  I wasn't thinking about the `q!...!` constructs, and the parser is not flexible enough to handle that (the entire PGML block is tokenized before anything is processed, but since the character used for the `q` can change, PGML can't tokenize the start and end delimiters because they aren't known up front, and we don't want to always match `q(.).*?\g{-1}`, as that would also be done *outside* of `[@...@]` where it shouldn't be (it also makes handling the token splitting a bit more complicated).

So perhaps the best idea is to not balance quoted in `[@...@]`.  That would mean removing `balance=>qr/[\'\"]/,` from

https://github.com/openwebwork/pg/blob/99fd91fd4a5d5d896399aab708abba210df65bb9/macros/PGML.pl#L478-L480

That means ``[@ "@]" @]`` will now end at the first `@]`, but that is not a very likely case.  You can also remove `parseComments=>1,` and  `parseSubstitutions=>1,` if you want to prevent that processing within the `[@ ... @]`.  The first may be useful for commenting out parts of a command.

**Comment by Alex-Jordan** [comment]
Thanks Davide, I see how that would work.

Just checking: is there a solution to this where `q` and `qq` are quote tokens without the additional character? And then each of the methods associated with a quote token or balancing would be modified to behave a little differently when the quote token is one of these?

**Comment by dpvc** [comment]
I've added a PR that handles the `q`, `qq`, and `qr` quoted string operators by looking through the token list by hand for the close delimiter and modifying the token list as it goes, rather than using the main token loop.  So this should handle your examples above, as well as allowing regular string quote balancing.

**Comment by drgrice1** [comment]
I am seeing a lot of warnings  with this right now.  You probably are seeing these too though.

The attached problem has a basic table.  I note a difference in the resulting displayed table from the develop branch version of niceTables.pl.  That is a border on the bottom of the `thead` section that wasn't there before.
[niceTablesTest.pg.txt](https://github.com/openwebwork/pg/files/10352598/niceTablesTest.pg.txt)

On POD, I recommend removing all of the `#` characters that seem to be used as a separator.  The alignment looks good on its own, and those characters just add noise.

Also, I recommend not using any tab characters in POD.  Tab characters are taken verbatim in POD and are assumed to be on 8-column boundaries.  That usually doesn't end up looking very good in the generated output from POD formatting programs.

Please remove the multiple empty lines that are in the POD.  A single empty line before and after any POD directive (lines starting with `=`) is required.  Anything more than makes POD take up more space than it needs to in the code.  POD formatting utilities remove all of those lines from the generated POD in any case.  perltidy removes multiple empty lines outside of POD, but perltidy won't touch anything inside POD.

Useful utilities for POD are `podchecker`  and `perldoc` which are included in most perl distributions by default.  You can also use the `bin/dev_scripts/generate-ww-pg-pod.pl` script from webwork2.  Example usage:
```bash
./bin/dev_scripts/generate-ww-pg-pod.pl -o ~/pod
```

**Comment by Alex-Jordan** [comment]
These comments sound like they apply to the nice tables PR, not the PGML
quotes PR. Is that right?

**Comment by drgrice1** [comment]
Yes they do.  Hmm, I seem to have to many tabs open in my browser.

**Comment by drgrice1** [comment]
I moved the comment there.  Sorry.

**Comment by pstaabp** [comment]
This is fixed with openwebwork/PG#760
